File: learn_algorithom/算法导论/15_chapter_dynamic_programming/15.4_lcs/C++/lcs.cpp
#include <iostream>
#include <iomanip>
#include <string>

using namespace std;

const int LEFT_UP = 0;
const int LEFT = 1;
const int UP = 2; 

void LCS_LENGTH(int **b, int **c, string X, string Y);
void PRINT_LCS(int **b, string X, int m, int n);

int main()
{
    string X = "belong", Y = "cnblogs";
    int m = X.length();
    int n = Y.length();
    int i, j;

    int **b = new int*[m], **c = new int*[m+1];

    // initialize b, c
    for (i=0; i<m; i++)
        b[i] = new int[n];
    for (j=0; j<=m; j++)
        c[j] = new int[n+1];

    LCS_LENGTH(b, c, X, Y);
    PRINT_LCS(b, X, m, n);
    cout << endl;

    // delete b, c
    for (i=0; i<m; i++)
        delete [] b[i];
    for (j=0; j<=m; j++)
        delete [] c[j];
}

void LCS_LENGTH(int **b, int **c, string X, string Y)
{
    using namespace std;
    int i,j;
    int m = X.length();
    int n = Y.length();

    for (i=0; i<=m; i++)
        c[i][0] = 0;
    for (j=0; j<=n; j++)
        c[0][j] = 0;

    for (i=1; i<=m; i++)
    {
        for (j=1; j<=n; j++)
        {
            if (X[i-1] == Y[j-1])
            {
                c[i][j] = c[i-1][j-1] + 1;
                b[i-1][j-1] =  LEFT_UP;
            } 
            else if(c[i-1][j] >= c[i][j-1]) 
            {
                c[i][j] = c[i-1][j];
                b[i-1][j-1] = UP;
            } else 
            {
                c[i][j] = c[i][j-1];
                b[i-1][j-1] = LEFT;
            }
        }
    }

    // print c
    //for (i=0; i<=m; i++)
    //{
        //for (j=0; j<=n; j++)
        //{
           //cout << setw(5) << c[i][j] ;
        //}
        //cout << endl;
    //}
        
    // print b
    //for (i=0; i<m; i++)
    //{
        //for (j=0; j<n; j++)
        //{
           //cout << setw(5) << b[i][j] ;
        //}
        //cout << endl;
    //}
}

void PRINT_LCS(int **b, string X, int m, int n)
{
    if (m==0 || n==0)
    {
        return;
    }
    else
    {
        if (b[m-1][n-1]==LEFT_UP)
        {
            PRINT_LCS(b, X, m-1, n-1);
            cout << X[m-1] << " ";
        }
        else
        {
            if (b[m-1][n-1] == UP)
            {
                PRINT_LCS(b, X, m-1, n);
            }
            else 
            {
                PRINT_LCS(b, X, m, n-1);
            }

        }
    }
}

File: learn_algorithom/算法导论/15_chapter_dynamic_programming/15.4_lcs/lcs.py


def LCS_LENGTH(x, y):

    n, m = len(x), len(y)
    c = [[None] * (m+1) for _ in range(n+1)]
    z = [[None] * m for _ in range(n)]

    for i in range(n+1):
        c[i][0] = 0
    for j in range(m+1):
        c[0][j] = 0

    for i in range(1, n+1):
        for j in range(1, m+1):
           
            if x[i-1] == y[j-1]:
                # 注意到此时lcs的长度加1，因此只需要找到 `left_up` 状态的就可以得到lcs
                c[i][j] = c[i-1][j-1] + 1
                z[i-1][j-1] = 'left_up'
            else:
                if c[i-1][j] > c[i][j-1]:
                    c[i][j] = c[i-1][j]
                    z[i-1][j-1] = 'up'
                else:
                    c[i][j] = c[i][j-1]
                    z[i-1][j-1] = 'left'
    return c, z

def LCS(x, y):

    c, z = LCS_LENGTH(x, y)

    # 从z中构造出子序列
    n, m = len(x), len(y)
    lcs = []

    i, j = n-1, m-1

    while i >= 0 and j >= 0:

        if z[i][j] == 'up':
            i -= 1
        if z[i][j] == 'left':
            j -= 1
        if z[i][j] == 'left_up':
            lcs.insert(0, x[i])
            i -= 1
            j -= 1
    return lcs

if __name__== "__main__":

    def test():
        x = 'belong'
        y = 'cnblogs'

        c, z = LCS_LENGTH(x, y)
        
        for _ in z:
            print(_)
        
        for _ in c:
            print(_)

        print(LCS(x, y))
    test()


